### ISSUE TITLE
Multiple users/codes

### ISSUE THREAD

**Issue by BoneheadFraggle** [thanksmister/android-mqtt-alarm-panel#65]
I really like this app and it would be a great feature to be able to add multiple users.
I read https://github.com/thanksmister/android-mqtt-alarm-panel/issues/34 but I'd like another approach:

I'd like to be able to add multiple users in MQTT Alarm Panel. Name and code would be enough - and  then publish a message which user that armed/disarmed. MQTT Alarm Panel would still work as today, which would secure the compatibility with HA Alarm Panel.

This way I could keep the actual code secret from my kids - but let them disarm the alarm the alarm with thier own code.

Passing a message which user that armed/disarmed would also open the possibility to automate actions.

Edit: I just realized that the exact same feature was suggested in Androidthings:
https://github.com/thanksmister/androidthings-mqtt-alarm-panel/issues/21

**Comment by lpt2007** [comment]
1+  vote to that suggestion

**Comment by nicolas3333** [comment]
Application works like a charm and I've got it integrated with Jeedom. I would also very much value this multi-user feature added and also to get the Telegram notification to say 'Paul disarmed the alarm' or 'Joe disarmed the alarm'.

**Comment by aaamoeder** [comment]
+1

**Comment by 5and0** [comment]
+1

Multi user codes is my last wish for my HA setup

**Comment by jaburges** [comment]
OMG this could be the missing link! 
I have 1 HUGE issue with all of my automations, when guests (not tracked with device_tracker) leave.
Do they all leave, does one stay - if they can arm the alarm with their code I KNOW that group of people (or person) is leaving and can arm the house. 
(the schlage lock the door from the outside has proven not to be overly reliable as a method of ensuring when guest leave)

**Comment by thedeejay** [comment]
+1
Either multi user in setup or something that integrates with bwalarm's users.
DISARM {"code": "0000"}
https://community.home-assistant.io/t/bwalarm-akasma74-edition

**Comment by fludstank** [comment]
Would be all for this.
Even if it was just to have the ability to include the incorrect codes in the payload (so that node-red could trigger different automations fror example a panic button situation).

**Comment by derekatkins** [comment]
I started to look at this, but got stuck pretty quickly just trying to add storage for a list of Name/Code mappings.  My plan was to just have a list in the app and if any of the codes from the list were provided then it would send the MQTT code over.  I'll have to wait until I have more time to dig deep into learning Kotlin.

**Comment by thanksmister** [comment]
The issue is, there is no way to parse out the codes from the payload using Home Assistant MQTT Alarm Panel.  If you try to send codes, the MQTT Alarm Panel will no longer respond to the MQTT messages.   You would have to parse the MQTT message and then handle the MQTT Alarm within HA.   This breaks the functionality of controlling the MQTT Alarm with the application.    I agree its a nice to have feature, but it would be best if the MQTT Alarm for HA would support this first.

**Comment by derekatkins** [comment]
What I meant was, let's assume the HA code is '1234'.  In the app I would have a list:
"Mom/Dad" 2345
"Baby sitter" 3456
...
Then if any of 2345 or 3456 were entered into the panel, it would send '1234' (or "disarm") to HA.  So HA would be completely unaware of the different codes being used.

Yes, I agree it would be nicer for HA to support multiple codes first.  I suspect that is "harder" to do than faking it in the app -- although it would be nice if there were a way to centrally control and configure multiple instances of the App (I currently have 2 panels, and I'm thinking of adding like 2 more -- so it would be nice to centrally control them all).

**Comment by thanksmister** [comment]
The message expected by the MQTT Alarm is very specific, you cannot just add extra parameter to the message.  MQTT Alarm component does not support sending codes.  You can only send a sing value such as DISARM as the payload.  You cannot also send a code.

**Comment by derekatkins** [comment]
Right, which is why I was thinking that I'd just maintain a list of valid codes in the App and send "disarm" on any valid code.  We'd lose the knowledge of which code was used, but it would at least separate out the master code (to get to the settings) from the alarm code(s).

**Comment by thanksmister** [comment]
Yeah, this is an application solution, but at this time, we cannot easily make it work on HA server.  :(

**Comment by derekatkins** [comment]
Yeah, until HA actually supports MQTT code(s), you're right that there isn't a good way to do that.  But we can "fake it" in the android app for now -- if only I could figure out how to serialize a listOf String,Int tuples .  lol.

**Comment by thanksmister** [comment]
You can use a hashmap Map<String, Int>?

**Comment by derekatkins** [comment]
I haven't been able to figure out the KT code (and gradle modifications) to convert a HashMap<String,Int> to a JSON string and vice-versa (in order to store it into and load it from a preference).  I haven't even started looking at a new pop-up page to edit the HashMap list!  :-(

**Comment by thanksmister** [comment]
You can create a list of pojos and convert it to a json string using gson.   So make like class for  User with properties and then make a ArrayList of User objects. Then you can use gson to make it json string, the reverse, gson to list of User objects.

**Comment by thanksmister** [comment]
My suggestion is, you have one setting for the main code.   Just add additional fields for up to 3 more optional code values.  These optional ones only unarm the alarm, they do not get into the settings.   Just make simple text field preferences that accept the code, then verify they are all numbers and only 4 digits before allowing to save.    So in the preferences, you have code (main) alarm code.  Then code 1, code 2, code 3.

**Comment by derekatkins** [comment]
That is certainly easier than a list of String/Int pairs ;)

**Comment by thanksmister** [comment]
Yeah, let's start out with simple feature, then we can modify it later.   For instance, we could add a mqtt message when the code is entered that notifies HA which code was entered.  You only have to keep track of which code you assign to which user.  It also looks like HA has added a code_template for sending the code over MQTT:

https://www.home-assistant.io/integrations/manual_mqtt/#code_template

```code_templatestring(exclusive)
If defined, returns a code to enable or disable the alarm in the frontend; an empty string disables checking the code. Inside the template, the variables from_state and to_state identify the current and desired state. Only one of code and code_template can be specified.```

From what I am reading, if you send the code template empty, then it works as normal, you can disarm the alarm without sending a code, which is what the app does now.  If we send a code in the template, its used as the code to disarm the alarm. Not sure if you could get this to support multiple codes.    We'd have to have an option to "Send code with disarm" so that its backwards compatible with users that have no code_template in their HA configuration or risk breaking everyone's alarm setup. 👍 

The other thing I am looking at is discovery, using the MQTT Alarm Control Panel  
https://www.home-assistant.io/integrations/alarm_control_panel.mqtt/ so that the app acts like a real MQTT alarm control panel and everything is just "discovered" by HA.

**Comment by derekatkins** [comment]
> Yeah, let's start out with simple feature, then we can modify it later.

Simple is always better.  I will look into that.
 
> https://www.home-assistant.io/integrations/manual_mqtt/#code_template

My reading of this is that you can have a different code to go from disarmed -> arm_home, disarmed -> arm_away, arm_home -> disarmed, and arm_away -> disarmed.  I don't really think that's what anybody wants (or needs).  Maybe I just don't understand what the code_template is really used for, but it doesn't *sound* useful to me.
 
> The other thing I am looking at is discovery, using the MQTT Alarm Control Panel
> https://www.home-assistant.io/integrations/alarm_control_panel.mqtt/ so that the app acts like a real MQTT alarm control panel and everything is just "discovered" by HA.

This would be really nice, provided you had a way to auto-synchronize multiple instances of the app on multiple tablets connected to the same HA/MQTT.  I think it would even be fine to declare one as the Main Panel that control everything, and the other as Subordinate Panels that take their configuration from the Main Panel / HA.

**Comment by thanksmister** [comment]
I thought the code_template was to send the code to MQTT, then the template would parse out the code and disarm the alarm.  This would be the way to actually send the alarm to HA and have it work with the alarm.   But without testing or seeing it in action, I am not sure how it works.   

It seem as though the discovery is using the unique ID of each device.  So each device would be treated as a separate alarm panel?   Again, really need to see how this works, also it appears that the application would have to be a registered device with home assistant for discovery to work.

**Comment by derekatkins** [comment]
My understanding of the Mqtt-Alarm (not manual_mqtt) is that it's a way to integrate HA into an actual alarm system  For example if you had a Honeywell alarm system and had an MQTT interface to it, you could control the alarm from HA.  That's not what I think we want to do here -- we're building an alarm in HA and want an interface to control it.

Having said that, I think it would be grand if there were more "discovery" in the app, where the App could learn the HA manual_mqtt configuration and therefore just be a code-input and state display for HA.  I.e., it would be nice to be able to move more of the logic into HA.  But that would certainly require more HA development (and coordination).

**Comment by thanksmister** [comment]
I think discovery works the opposite, HA discovers the device, and then the device tells HA how to control it.   From the docs 

> The discovery of MQTT devices will enable one to use MQTT devices with only minimal configuration effort on the side of Home Assistant. The configuration is done on the device itself and the topic used by the device. 

The application must adhere to how HA "discovers" devices and what they allow for communication.  The device will tell HA what topics its going to use.  

https://www.home-assistant.io/docs/mqtt/discovery/

**Comment by thanksmister** [comment]
I mean, we could be talking about the same thing, the alarm panel will have to know about HA and vice versa, its just that users will not have to configure HA ahead of time, the device will tell HA what to configure for them.   This way, someone can install the app and then have HA discover it and its automatically working.    

I think in this case, codes being sent to HA are not important, but the alarm panel handles codes and security, like a real alarm panel would.  The HA server should be protected from guests and users.   Since I started developing apps for HA, I have leaned towards having HA as a server, not the interface everyone interacts with.  So having the code sent to HA is not important to me.
